### PR TITLE
Optimise conversion of dense data, rewriting decoding

### PR DESCRIPTION
**Pull request by jnothman**
This takes half of the loading time off for MNIST.

TODO:

* [x] raise BadNumericalValue as appropriate
* [x] fix normalisation in sparse cases. This is currently not tested. Unfortunately, pulling the cleaning code out for reuse makes it much slower, so we might need to duplicate code.

Ping @rth to do a better benchmark :)

### CODE REVIEW

**Comment by rth** [comment]
I can confirm. With the simple benchmark script below,

```py
from time import time

import arff
from scipy.io.arff import loadarff

fname = "mnist_784.arff"

t0 = time()

with open(fname, 'rt') as fh:
    data = arff.load(fh)

print('liac-arff: {:.1f} s'.format(time() - t0))


t0 = time()
data = loadarff(fname)

print('scipy-arff: {:.1f} s'.format(time() - t0))
```
</details>

On my laptop with SSD, I get,
```
liac-arff: 27.8 s
scipy-arff: 21.8 s
```
on a server with HDD,
```
liac-arff: 36.7 s
scipy-arff: 27.5 s
```

so performance wise this looks quite close to the scipy parser. Nice work!

**Comment by mfeurer** [comment]
I just checked the new code on my laptop (with an SSD):
```
/home/feurerm/miniconda/3-4.5.4/envs/smac3/bin/python3 /home/feurerm/sync_dir/Software/github_forks/liac-arff/test_things.py
liac-arff: 31.3 s
scipy-arff: 32.8 s
```

> This is now basically a rewrite of decoding. I wish I was more confident that we have thorough tests... I think there should probably be more testing of the write-read round trip (with each dense/sparse variant).

I see if I can check some more datasets from OpenML.org. In the long run I would love to actually compare against the behavior of WEKA, but I think I won't be able to do this until the end of the next week.

> I've not yet benchmarked the sparse case, or the case with quotes, but the dense case on NIST is better than I have seen it. 

Sparse and quotes are edge cases on OpenML. I don't think focusing on them in a first iteration is the way to go.

> I can only presume that this regex parser is faster than the previous _read_csv.

Yeah, I guess so. I only added that to cope with the fact that one can mix quotation marks in the arff files.

> Syntax errors are less specific than they used to be, and I still haven't understood the difference between all the classes of exception 

Cool, I'll check if this is also still compatible with the OpenML package.

> (or why we need so many).

Historic reasons.

> please advise if there is anything I could help to clarify in order to proceed with review.

I'll start reviewing in a few minutes. I might only finish on Monday, though.

> Although basically, if we don't believe the changes in tests could be detrimental to OpenML support, I would be keen to see this version adopted in Scikit-learn.

Most datasets on OpenML are rather simple and the previous version of liac-arff was able to handle most use cases so far. I believe that the new tests improve the package, I will see if I can find the time to add some more.

**Comment by mfeurer** [comment]
I though the best way to check the parser is to compare it against the old parser on all OpenML datasets:
```
import arff_old
import arff_new
import os
import openml
import requests


datasets = openml.datasets.list_datasets()
for key in datasets:

    print(key)
    url = 'https://www.openml.org/data/download/%d' % key
    arff_string = requests.get(url).content.decode('utf8')

    failure = False
    try:
        d_1 = arff_old.loads(arff_string)
    except Exception as e:
        print('Failure to load arff with non-regex version.')
        print(e)
        failure = True
    try:
        d_2 = arff_new.loads(arff_string)
    except Exception as e:
        print('Failure to load arff with regex version.')
        print(e)
        failure = True
    if failure:
        continue
    for arff_key in d_1:
        if d_1[arff_key] != d_2[arff_key]:
            print(arff_key)
```
Tests are running...

**Comment by jnothman** [comment]
Any failures?

**Comment by mfeurer** [comment]
Not yet, but my script was wrong so it didn't work over the weekend. Here's a correct version:
```
import arff_old
import arff_new
import os
import openml
import requests

openml.config.cache_directory = '/tmp/openml/'


datasets = openml.datasets.list_datasets()
for key in datasets:
    print(key)
    dataset = openml.datasets.get_dataset(key)

    with open('/tmp/openml/org/openml/www/datasets/%d/dataset.arff' % key) as fh:
        arff_string = fh.read()

    failure = False
    try:
        d_1 = arff_old.loads(arff_string)
    except Exception as e:
        print('Failure to load arff with non-regex version.')
        print(e)
        failure = True
    try:
        d_2 = arff_new.loads(arff_string)
    except Exception as e:
        print('Failure to load arff with regex version.')
        print(e)
        failure = True
    if failure:
        continue
    for arff_key in d_1:
        if d_1[arff_key] != d_2[arff_key]:
            print(arff_key)
``` 
I'll check one more after lunch and let you know.

**Comment by jnothman** [comment]
It looks like there was a bug in encoding that involved double-quoting, or expecting some input to be pre-quoted, or something? I don't really get how it was broken before, except that the expected output in `tests/test_dumps_escape.py` didn't match my expectations.

File: tests/test_dumps_escape.py [hunk 1]
```diff
@@ -36,7 +36,7 @@
 lorem,b,'this is a valid statement with an \\% symbol',\'\\"b\\"\'
 lorem2,d,'this is a valid statement',\'\\"d\\"\'
 lorem3,c,'this is a valid statement with double quotes included \\"\\"\\"\\"\\"\\"\\"! ',\'\\"c\\"\'
-lorem4,a,'this is a valid statement with singlequotes included \\\' lol \\\'! ',\'\\"a\\"\'
```
Comment (jnothman) [inline tests/test_dumps_escape.py]: Basically, this should have been `"\\\\\\\'"` if it was encoding `"\\\'"`, not `"\\\'"`...

```py
>>> s = "\\\'"
>>> print(s)
\'
>>> print(arff.encode_string(s))
'\\\''
>>> arff.encode_string(s)
"'\\\\\\''"
```

Comment (mfeurer) [reply tests/test_dumps_escape.py]: Do you mean there's an error here? I tested to read and write the following arff and the results are consistent with WEKA:
```
@RELATION string_escape

@ATTRIBUTE string STRING

@DATA

"\'"
"\\'"
"\\\'"
"\\\\'"
"\\\\\'"
"\\\\\\'"
```
Output:
```
@RELATION string_escape

@ATTRIBUTE string STRING

@DATA
'\''
'\\\''
'\\\''
'\\\\\''
'\\\\\''
'\\\\\\\''
```
WEKA
```
@relation string_escape

@attribute string string

@data
'\''
'\\\''
'\\\''
'\\\\\''
'\\\\\''
'\\\\\\\''
```

**Comment by mfeurer** [comment]
Thanks for the additional tests. I'd be happy to merge this now, is there anything open from your side @jnothman ?

File: tests/test_dumps_escape.py (hunk 1 above)
Comment (jnothman) [reply tests/test_dumps_escape.py]: Yes, there is an error in the functionality in master, required for this test to pass. Now the test and implementation are fixed, imo.

**Comment by jnothman** [comment]
Nothing remaining, except that the exception types are a bit under specified, and some may now be unused. 

I took the liberty to modify CHANGES.rst; I suppose you agree this is worth a release.

**Comment by mfeurer** [comment]
Okay, great! I think weeding through the exceptions should be in a separate PR. I'll push a release to pypi.

**Pull request merged**